### PR TITLE
CP-26098: Upgrade xapi to use PPX-based v6d interface

### PR DESCRIPTION
**Pull request by minishrink**
This PR is the second of three which port the licensing daemon to PPX:
1. xcp-idl: [port V6d interface to use PPX](https://github.com/xapi-project/xcp-idl/pull/186)
2. xapi: update references to v6 interface
3. v6d: port daemons to ppx

This PR in particular updates xapi to use the following:
- ppx-based error handling (defined in v6 interface)
- new edition type (defined in v6 interface)
- some formatting changes for readability, as well as whitespace removal and ocp-indent

Note: CP-26098 is part of REQ-417

---
This change is [Reviewable](https://reviewable.io/reviews/xapi-project/xen-api/3391)

### CODE REVIEW

**Comment by minishrink** [comment]
I found myself making a lot of formatting changes, so those might come in a separate PR.

File: ocaml/xapi/cli_operations.ml [hunk 1]
```diff
@@ -208,14 +208,19 @@ let host_license_of_r host_r editions =
...
-  let edition_short = List.hd
-      (List.filter_map (fun (a, (_, b, _)) -> if a = edition then Some b else None) editions) in
+  let edition_str = host_r.API.host_edition in
+  let edition_short =
+    editions
+    |> List.filter_map
+      V6_interface.(fun ed -> if ed.ed_title = edition_str
+                     then Some ed.ed_code
+                     else None)
+    |> List.hd in
```
Comment (robhoes) [inline ocaml/xapi/cli_operations.ml]: We should really get rid of this List.hd...

File: ocaml/xapi/cli_operations.ml [hunk 2]
```diff
@@ -208,14 +208,19 @@ let host_license_of_r host_r editions =
...
+     V6_interface.(fun ed -> if ed.ed_title = edition_str
+                                then Some ed.ed_code
+                                else None)
+    |> List.hd in
   {
     hostname = host_r.API.host_hostname;
     uuid = host_r.API.host_uuid;
     rstr = rstr;
-    edition = edition;
+    edition = edition_str;
```
Comment (robhoes) [inline ocaml/xapi/cli_operations.ml]: What's the benefit of the rename?

File: ocaml/xapi/cli_operations.ml [hunk 3]
```diff
@@ -3249,16 +3254,15 @@ let with_license_server_changes printer rpc session_id params hosts f =
 
 let host_apply_edition printer rpc session_id params =
   let host =
-    if List.mem_assoc "host-uuid" params then
-      Client.Host.get_by_uuid rpc session_id (List.assoc "host-uuid" params)
-    else
-      get_host_from_session rpc session_id in
+    if List.mem_assoc "host-uuid" params
+    then Client.Host.get_by_uuid rpc session_id (List.assoc "host-uuid" params)
```
Comment (robhoes) [inline ocaml/xapi/cli_operations.ml]: This seems an unnecessary change?

File: ocaml/xapi/xapi_host.ml [hunk 4]
```diff
@@ -1339,16 +1339,18 @@ let apply_edition_internal  ~__context ~host ~edition ~additional =
   let params =
     ("current_edition", current_edition) ::
     license_server @ current_license_params @ additional in
+  let edition_str = edition in
+  (* rename edition variable to prevent namespace clashes with V6_interface *)
```
Comment (robhoes) [inline ocaml/xapi/xapi_host.ml]: Where then? I don't see a `let edition` or something like that in there?

Comment (mseri) [reply ocaml/xapi/xapi_host.ml]: It could be renamed in the function definition itself: https://github.com/xapi-project/xen-api/pull/3391/files/634cfc05e57fb3fbaab6312554d0c34dab103857..733531584953d17ce23901200b4f06e6db122086#diff-126a0619d5403393ed49475b5dd96aa4R1326

Comment (robhoes) [reply ocaml/xapi/xapi_host.ml]: It can, but why?

File: ocaml/xapi/cli_operations.ml (hunk 1 above)
Comment (minishrink) [reply ocaml/xapi/cli_operations.ml]: Replace it with a safer List.hd using options?

File: ocaml/xapi/cli_operations.ml (hunk 2 above)
Comment (minishrink) [reply ocaml/xapi/cli_operations.ml]: I think I did that to keep some consistency with the renaming in v6d, but you're right in that it serves no purpose here so I'll remove it.

File: ocaml/xapi/cli_operations.ml (hunk 1 above)
Comment (robhoes) [reply ocaml/xapi/cli_operations.ml]: It's usually better to use a match to get the first element of a list, and explicitly handle the `[]` case. The exception raised by `List.hd` in case the list is empty is often not what you want.

File: ocaml/xapi/xapi_host.ml (hunk 4 above)
Comment (mseri) [reply ocaml/xapi/xapi_host.ml]: You are right, I thought we were in the same situation as in `v6d` but here it's not needed

File: ocaml/xapi/cli_operations.ml (hunk 2 above)
Comment (minishrink) [reply ocaml/xapi/cli_operations.ml]: Satisfying the type system, which will otherwise recognise `edition` as a generated RPC type.

Comment (minishrink) [reply ocaml/xapi/cli_operations.ml]: Another way to deal with this would be to change the type name from edition to edition_repr or edition_rec, edition_type, or something

File: ocaml/xapi/xapi_host.ml (hunk 4 above)
Comment (minishrink) [reply ocaml/xapi/xapi_host.ml]: As I understand it, prior to locally opening `V6_interface`, `edition` is interpreted as a string. Within the V6_interface namespace, it is interpreted as an RPC type instead of a string, which upsets the compiler.

Which do you think is the best solution?
1. Rename the `edition` type in the V6 interface to avoid clashes
2. Rename to `edition_str` within the function definition (this requires updating elsewhere as well)
3. Local rename for this small snippet of code (used in this PR)

File: ocaml/xapi/cli_operations.ml (hunk 1 above)
Comment (minishrink) [reply ocaml/xapi/cli_operations.ml]: As the `edition_short` value is used in the host license record, should we handle a failure with `free`, `""`, or by raising an error?

File: ocaml/xapi/cli_operations.ml (hunk 3 above)
Comment (minishrink) [reply ocaml/xapi/cli_operations.ml]: I may have gone on a slight spree in order to make some code more readable (have since moved that to a separate branch), though if/then/else just comes down to preference. Should I remove it, or just bundle it into a separate PR?

File: ocaml/xapi/xapi_host.ml (hunk 4 above)
Comment (robhoes) [reply ocaml/xapi/xapi_host.ml]: Or 4: don't locally open `V6_interface` anymore, but just use `V6_interface.`.

Comment (minishrink) [reply ocaml/xapi/xapi_host.ml]: Also works!

File: ocaml/xapi/cli_operations.ml (hunk 3 above)
Comment (robhoes) [reply ocaml/xapi/cli_operations.ml]: I happen to think that the original code is more readable ;)

Generally speaking, we should avoid making arbitrary style changes unless your changing those lines for another reason as well, or the new version is obviously Much Better™.

File: ocaml/xapi/xapi_host.ml [hunk 5]
```diff
@@ -1400,7 +1400,7 @@ let apply_edition ~__context ~host ~edition ~force =
     raise (Api_errors.Server_error (Api_errors.ha_is_enabled, []))
   else
     let additional = if force then ["force", "true"] else [] in
-    apply_edition_internal ~__context ~host ~edition ~additional
+    apply_edition_internal ~__context ~host ~edition:edition ~additional
```
Comment (robhoes) [inline ocaml/xapi/xapi_host.ml]: That shouldn't be needed.

Comment (minishrink) [reply ocaml/xapi/xapi_host.ml]: Whoops, holdover from removing `edition_str`.
